keps/sig-apps/20181220-statefulset-volume-expansion.md

```diff
@@ -0,0 +1,103 @@
+
+It would be easier and expected to allow for changes to storage requests in the `volumeClaimTemplates` component of a StatefulSet to propagate to all associated persistent volume claims.
+
+Relevant Issues:
+
+* https://github.com/kubernetes/kubernetes/issues/71477
+* https://github.com/kubernetes/kubernetes/issues/72198
+
+### Goals
+
+Allow for increases to storage requests in the `volumeClaimTemplates` component of a StatefulSet to propagate to all associated persistent volume claims.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+The apiserver will allow for increases to storage requests in the `volumeClaimTemplates` component of a StatefulSet. Additionally, for each volumeClaimTemplate being expanded, it will be necessary to check if its associated StorageClass has volume expansion enabled. This will be achieved by using an admission controller to check if the associated StorageClasses have `allowVolumeExpansion = true`.
+
+During the StatefulSet update process, the StatefulSet controller will detect an update to a `volumeClaimTemplate` by comparing the updated and current revision of the StatefulSet. This requires the `VolumeClaimTemplates` component of the StatefulSet to be recorded in the StatefulSet's `ControllerRevision` object. 
+
+While updating a pod, the StatefulSet controller will update a referenced persistent volume claim object if its storage request in the associated `volumeClaimTemplate` has been increased.
+
+Not all volumes support online control-plane expansion so this design aims to support both online and offline control-plane volume expansion as follows.
+
+Prerequisite: The StatefulSet controller must be able to determine if a volume supports online expansion.
+
+To accomplish this a new method `RequiresOfflineResize` will be added to the `ExpandableVolumePlugin` interface. Through this it will possible to determine which in-tree volume plugins support online expansion. 
+
+For volumes provisioned through the CSI, an optional boolean field `SupportsOnlineExpansion` will be added into the `CSIDriverSpec` field of the `CSIDriver` object. The `CSIDriver` object stores information about a CSI volume driver deployed on the cluster. If the `CSIDriver` object or the `SupportsOnlineExpansion` field is not registered for a particular volume provisioned through the CSI, the statefulset controller will assume the volume does not support online expansion. Adding this functionality mandates an API change for the mentioned objects.
```

Comment:
I don't like the statefulset controller having to look at this info.  This feels complex.  What other alternatives have been considered.

Reply:
We have considered a couple other alternatives in this [google doc](https://docs.google.com/document/d/18ltkK6VtoHxwA3qcurU2k0rhoowBEpLVqQkZYdiNKz4/edit?usp=sharing). 

Summary:

The primary alternative was expanding all pvcs offline right after the pod is terminated (regardless of whether they support online expansion). 

A workaround was expanding PVCs online if possible, otherwise just continuing with the update and terminating the pod while expecting user to expand the pvc themselves after which the pod would be recreated. (adding an event etc). 

The reason we would like to expand the PVCs online is so if a user sees that the replicas in a statefulset are running out of space it should be possible for the user to just expand those volumes without without having each pod be offline during the duration of the volume expansion(which could take a while).  At the same time we would not like to abandon the offline only expansion case thus the strategy in the kep seemed optimal.

Reply:
@SidakM but even if we expand volumes ONLINE we are still going to restart the pod. How does that minimizes downtime vs terminating the pod before?

Reply:
When we expand online, the pod can stay running while the controller performs the resize.

If we terminate the pod before, it will not be running while the controller performs the resize operation.

Reply:
Agree that having to handle offline/online controller expansion differently complicates the design.  I mainly see handling online expansion specially as an optimization for pod downtime, because in the end, we still need to restart the pod.  Can we simplify the design initially and handle everything one way?

Reply:
We can expand only while the pod is terminated if we want to keep things simple. But I think it is definitely advantageous to expand volumes online if not now but in the future to minimize downtime and eventually not having to restart at all. 

Given this do we want to roll with expanding only offline @msau42 @gnufied

Reply:
I am okay going only with offline in this alpha release.